src/libraries/Microsoft.Extensions.Configuration/src/ConfigurationKeyComparer.cs

```diff
@@ -29,49 +29,80 @@ public class ConfigurationKeyComparer : IComparer<string>
         /// <returns>Less than 0 if x is less than y, 0 if x is equal to y and greater than 0 if x is greater than y.</returns>
         public int Compare(string? x, string? y)
         {
-            string[] xParts = x?.Split(_keyDelimiterArray, StringSplitOptions.RemoveEmptyEntries) ?? Array.Empty<string>();
-            string[] yParts = y?.Split(_keyDelimiterArray, StringSplitOptions.RemoveEmptyEntries) ?? Array.Empty<string>();
+            ReadOnlySpan<char> xSpan = x.AsSpan();
+            ReadOnlySpan<char> ySpan = y.AsSpan();
+
+            xSpan = SkipAheadOnDelimiter(xSpan);
+            ySpan = SkipAheadOnDelimiter(ySpan);
 
             // Compare each part until we get two parts that are not equal
-            for (int i = 0; i < Math.Min(xParts.Length, yParts.Length); i++)
+            while (!xSpan.IsEmpty && !ySpan.IsEmpty)
             {
-                x = xParts[i];
-                y = yParts[i];
+                int xSegmentLength = xSpan.IndexOf(s_keyDelimiter);
+                if (xSegmentLength < 0)
+                {
+                    xSegmentLength = xSpan.Length;
+                }
+
+                int ySegmentLength = ySpan.IndexOf(s_keyDelimiter);
+                if (ySegmentLength < 0)
+                {
+                    ySegmentLength = ySpan.Length;
+                }
+
+                int compareResult = Compare(xSpan.Slice(0, xSegmentLength), ySpan.Slice(0, ySegmentLength));
+                if (compareResult != 0)
+                {
+                    return compareResult;
+                }
+
+                xSpan = xSpan.Slice(xSegmentLength);
```

Comment:
It's a little surprising that in the middle of the path, we know `xSpan.Slice(xSegmentLength)` will have a delimiter at the beginning, and then we slice it off in SkipAheadOnDelimiter.

I guess I would have expected this to check the result of IndexOf twice, once during the slice above, and once here when slicing for the next segment.

```C#
int xDelimiterIndex = xSpan.IndexOf(s_keyDelimiter);
int yDelimiterIndex = ySpan.IndexOf(s_keyDelimiter);

int compareResult = Compare(
    xDelimiterIndex == -1 ? xSpan : xSpan.Slice(0, xDelimiterIndex),
    yDelimiterIndex == -1 ? ySpan : ySpan.Slice(0, yDelimiterIndex));
if (compareResult != 0)
{
    return compareResult;
}

xSpan = xDelimiterIndex == -1 ? default :
    SkipAheadOnDelimiter(xSpan.Slice(xDelimiterIndex + 1));
ySpan = yDelimiterIndex == -1 ? default :
    SkipAheadOnDelimiter(ySpan.Slice(yDelimiterIndex + 1));
```